### PR TITLE
Major UI improvements

### PR DESCRIPTION
**Pull request by asaadmahmood**
#### Summary
Major UI improvements
- Includes revamp of Todo list.
- Opens assignee modal in a list
- Ability to change assignee after todo created
- Added todo description field
- Removed ask someone a todo from the out list.
- Added undo toasts for Complete and Remove.
- Added menu to allow users to do actions from the list.
- Added shortcuts to add new todo, edit todo, delete todo, assign, save, discard etc
- Moved In and out tabs into a list.
- Added empty state for todo list.

### CODE REVIEW

File: webapp/src/components/add_issue/add_issue.jsx [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        submit: PropTypes.func.isRequired,
+        theme: PropTypes.object.isRequired,
+        autocompleteUsers: PropTypes.func.isRequired,
+        openAssigneeModal: PropTypes.func.isRequired,
+        removeAssignee: PropTypes.func.isRequired,
+    };
+
+    shouldComponentUpdate() {
+        return true;
+    }
```
Comment (larkox) [inline webapp/src/components/add_issue/add_issue.jsx:37]: Is this needed?

File: webapp/src/components/add_issue/add_issue.jsx [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        removeAssignee();
+        closeAddBox();
+    };
+
+    toggleAssigneeModal = (value) => {
+        this.setState({ assigneeModal: value });
+    }
+
+    onKeyDown = (e) => {
+        if (e.key === 'Enter') {
```
Comment (larkox) [inline webapp/src/components/add_issue/add_issue.jsx:104]: Do you want to consider things like cmd+enter to add a new line?

File: webapp/src/components/assignee_modal/assignee_modal.jsx [hunk 3]
```diff
@@ -0,0 +1,147 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+
+import AutocompleteSelector from '../user_selector/autocomplete_selector.tsx';
+import Button from '../../widget/buttons/button';
+import IconButton from '../../widget/iconButton/iconButton';
+
+import CompassIcon from '../icons/compassIcons';
+
+const AssigneeModal = ({ visible, close, autocompleteUsers, theme, getAssignee, removeAssignee, removeEditingTodo, changeAssignee, editingTodo }) => {
```
Comment (larkox) [inline webapp/src/components/assignee_modal/assignee_modal.jsx:10]: Line too long. Can you put each argument in a new line?

File: webapp/src/components/assignee_modal/assignee_modal.jsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    const [assignee, setAssignee] = useState();
+
+    useEffect(() => {
+        function handleKeypress(e) {
+            if (e.key === 'Escape' && visible) {
```
Comment (larkox) [inline webapp/src/components/assignee_modal/assignee_modal.jsx:15]: If the component gets instantiated as not visible, this won't be executed at any moment, even if visible changes. If the component is instantiated as visible, but then changes, this would be still executed.

File: webapp/src/components/assignee_modal/assignee_modal.jsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        return () => {
+            document.removeEventListener('keyup', handleKeypress);
+        };
+    }, []);
+
+    if (!visible) {
+        return null;
+    }
+
+    const submit = () => {
```
Comment (larkox) [inline webapp/src/components/assignee_modal/assignee_modal.jsx:31]: Same comment about useCallback in this file.

File: webapp/src/components/assignee_modal/assignee_modal.jsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                <IconButton
+                    size='medium'
+                    style={style.closeIcon}
+                    onClick={closeModal}
+                    icon={<CompassIcon icon='close'/>}
+                />
+                <AutocompleteSelector
+                    id='send_to_user'
+                    loadOptions={autocompleteUsers}
+                    onSelected={(selected) => changeAssigneeDropdown(selected)}
```
Comment (larkox) [inline webapp/src/components/assignee_modal/assignee_modal.jsx:69]: Same comment as before on inline functions.

File: webapp/src/components/buttons/complete.jsx [hunk 7]
```diff
@@ -1,18 +1,34 @@
...
+        }, 1000);
+    };
+
     return (
         <button
-            className='btn btn-primary'
-            onClick={() => props.complete(props.issueId)}
-        >{'Done'}</button>
+            className={`todo-item__checkbox ${props.active ? 'todo-item__checkbox--active' : ''}`}
+            onClick={() => markAsDone()}
```
Comment (larkox) [inline webapp/src/components/buttons/complete.jsx:17]: No need for the arrow function. You can directly pass `markAsDone`.

File: webapp/src/components/buttons/complete.jsx [hunk 8]
```diff
@@ -1,18 +1,34 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 
+import CompassIcon from '../icons/compassIcons';
+
 const CompleteButton = (props) => {
+    const markAsDone = () => {
```
Comment (larkox) [inline webapp/src/components/buttons/complete.jsx:7]: useCallback

File: webapp/src/widget/menu.tsx [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+                            className='menu-cancel'
+                            onClick={this.onCancel}
+                        />
+                    </div>
+                </div>
+            </div>
+        );
+    }
+
+    private onCancel = () => {
```
Comment (larkox) [inline webapp/src/widget/menu.tsx:42]: The effect is not the same if we just remove the onClick property? Or the onClick property is needed?

File: webapp/src/widget/menuItem.tsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    text?: string,
+    shortcut?: string,
+}
+
+const MenuItem = (props: Props) => {
+    const { icon, shortcut, action, text } = props;
+
+    useEffect(() => {
+        function handleKeypress(e: {key: string}) {
+            if (e.key === shortcut) {
```
Comment (larkox) [inline webapp/src/widget/menuItem.tsx:18]: If shortcut changes during the lifetime of the component, this won't change. Is this expected?

File: webapp/src/widget/menuWrapper.tsx [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        if (e.key === 'Escape') {
+            close();
+        }
+
+        if (e.key === 'Tab') {
+            closeOnBlur(e);
+        }
+    };
+
+    const toggle = (e: React.MouseEvent<HTMLDivElement, MouseEvent>): void => {
```
Comment (larkox) [inline webapp/src/widget/menuWrapper.tsx:47]: More useCallback opportunities in this file.

File: webapp/src/components/todo_item/todo_item.jsx [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+        setHidden(false);
+    };
+
+    const undoCompleteTodo = () => {
+        clearTimeout(completeTimeout);
+        setHidden(false);
+        setDone(false);
+    };
+
+    const completeToast = () => {
```
Comment (larkox) [inline webapp/src/components/todo_item/todo_item.jsx:110]: More useCallback candidates.

**Review by larkox — CHANGES_REQUESTED** [review]
Awesome work! A few changes here and there, but looks great.

File: .vscode/extensions.json [hunk 13]
```diff
@@ -0,0 +1,6 @@
+{
```
[the remark this answers is not in the document]
Comment (asaadmahmood) [reply .vscode/extensions.json:1]: They are in the focalboard repo, so I'll keep them.

File: webapp/src/widget/menu.tsx (hunk 9 above)
Comment (asaadmahmood) [reply webapp/src/widget/menu.tsx:42]: I think its needed, that's how it done on boards too, we have a comment on it as well.

File: webapp/src/components/todo_item/todo_item.jsx [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+            issueId={issue.id}
+            markAsDone={() => setDone(true)}
+            completeToast={completeToast}
+        />
+    );
+
+    const removeTodo = useCallback(() => {
+        openTodoToast({icon: 'trash-can-outline', message: 'Todo deleted', undo: undoRemoveTodo});
+        setHidden(true);
+        removeTimeout = setTimeout(() => {
```
Comment (larkox) [inline webapp/src/components/todo_item/todo_item.jsx:129]: This won't work. You need to use a reference, like you did with completeTimeout.

The reason is the following:
- Every render, a new reference to removeTimeout is created.
- The value of removeTimeout (undefined) is used to create undoRemoveTodo.
- RemoveTodo sets a value to the removeTimeout variable, but this does not get reflected in the body of `undoRemoveTodo`, since it is already using old value (undefined).

If we use a reference:
- On every render, the reference to removeTimeout is the same (a reference to an object like `{current: something}`)
- undoCompleteTodo receives the value of removeTimeout (a reference to the object `{current: something}`)
- removeTodo assigns a new value not to removeTimeout, but to removeTimeout.current. Since we are working with references to objects here, the new value is ready to be referenced on `undoCompleteTodo`

I hope it is a bit clearer now.

File: webapp/src/components/todo_item/todo_item.jsx [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+
+    const completeToast = useCallback(() => {
+        openTodoToast({icon: 'check', message: 'Todo completed', undo: undoCompleteTodo});
+
+        setHidden(true);
+
+        completeTimeout.current = setTimeout(() => {
+            complete(issue.id);
+        }, 5000);
+    }, [complete, openTodoToast, issue, completeTimeout]);
```
Comment (larkox) [inline webapp/src/components/todo_item/todo_item.jsx:103]: completeTimeout is always the same between renders (because of how `useRef` works) so there is no need to have it in the dependencies.

File: webapp/src/widget/todo_toast/todo_toast.tsx [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+function TodoToast(props: Props): JSX.Element {
+    const {close, content} = props;
+
+    const closeToast = useCallback(close, [close]);
+    const undoTodo = useCallback(() => {
+        content.undo();
+        close();
+    }, [content.undo(), close]);
```
Comment (larkox) [inline webapp/src/widget/todo_toast/todo_toast.tsx:32]:
```suggestion
    }, [content.undo, close]);
```
You want to make sure the function is the same function, not the result of the function (which probably is void).

**Review by larkox — CHANGES_REQUESTED** [review]
Some minor changes, and also some comments still not addressed from my previous feedback.

There are still some potential improvements in terms of using `useCallback` or `useMemo`, but shouldn't be a big deal.

File: webapp/src/components/add_issue/add_issue.jsx (hunk 2 above)
Comment (asaadmahmood) [reply webapp/src/components/add_issue/add_issue.jsx:106]: Not at this point. Maybe as the next iteration we can add it in.

File: webapp/src/components/assignee_modal/assignee_modal.jsx (hunk 4 above)
Comment (asaadmahmood) [reply webapp/src/components/assignee_modal/assignee_modal.jsx:15]: Should be fixed if we pass visible as a parameter to useEffect.

**Review by larkox — APPROVED** [review]
LGTM.

File: server/plugin.go [hunk 17]
```diff
@@ -318,6 +327,108 @@ func (p *Plugin) handleList(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	var editRequest *editAPIRequest
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&editRequest); err != nil {
+		p.API.LogError("Unable to decode JSON err=" + err.Error())
+		p.handleErrorWithCode(w, http.StatusBadRequest, "Unable to decode JSON", err)
+		return
+	}
```
Comment (mickmister) [inline server/plugin.go:349]: We'll want to call `r.Body.Close()` after this block

File: server/plugin.go [hunk 18]
```diff
@@ -318,6 +327,108 @@ func (p *Plugin) handleList(w http.ResponseWriter, r *http.Request) {
...
+
+	var changeRequest *changeAssignmentAPIRequest
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&changeRequest); err != nil {
+		p.API.LogError("Unable to decode JSON err=" + err.Error())
+		p.handleErrorWithCode(w, http.StatusBadRequest, "Unable to decode JSON", err)
+		return
+	}
```
Comment (mickmister) [inline server/plugin.go:394]: Same concern here, we should call `r.Body.Close()` after this block

File: server/plugin.go [hunk 19]
```diff
@@ -318,6 +327,108 @@ func (p *Plugin) handleList(w http.ResponseWriter, r *http.Request) {
...
+
+	if changeRequest.SendTo == "" {
+		http.Error(w, "No user specified", http.StatusBadRequest)
+		return
+	}
+
+	receiver, appErr := p.API.GetUserByUsername(changeRequest.SendTo)
+	if appErr != nil {
+		p.API.LogError("username not valid, err=" + appErr.Error())
+		p.handleErrorWithCode(w, http.StatusInternalServerError, "Unable to find user", appErr)
```
Comment (mickmister) [inline server/plugin.go:404]:
```suggestion
		p.handleErrorWithCode(w, http.StatusNotFound, "Unable to find user", appErr)
```

File: webapp/src/components/add_issue/add_issue.jsx [hunk 20]
```diff
@@ -0,0 +1,332 @@
...
+        closeAddBox: PropTypes.func.isRequired,
+        submit: PropTypes.func.isRequired,
+        theme: PropTypes.object.isRequired,
+        autocompleteUsers: PropTypes.func.isRequired,
+        openAssigneeModal: PropTypes.func.isRequired,
+        removeAssignee: PropTypes.func.isRequired,
+    };
+
+    shouldComponentUpdate() {
+        return true;
```
Comment (mickmister) [inline webapp/src/components/add_issue/add_issue.jsx:36]: This is not good practice. WHy is this needed?

File: webapp/src/components/add_issue/add_issue.jsx [hunk 21]
```diff
@@ -0,0 +1,332 @@
...
+            previewMarkdown: false,
+            assigneeModal: false,
+        };
+    }
+
+    static getDerivedStateFromProps(props, state) {
+        if (props.visible && state.message == null) {
+            return {message: props.message};
+        }
+        if (!props.visible && (state.message != null || state.sendTo != null)) {
```
Comment (mickmister) [inline webapp/src/components/add_issue/add_issue.jsx:56]: Any reason why this won't work?

```suggestion
        if (!props.visible && (state.message || state.sendTo)) {
```

File: webapp/src/components/add_issue/add_issue.jsx [hunk 22]
```diff
@@ -0,0 +1,332 @@
...
+            description: '',
+            sendTo: null,
+            attachToThread: false,
+            previewMarkdown: false,
+            assigneeModal: false,
+        };
+    }
+
+    static getDerivedStateFromProps(props, state) {
+        if (props.visible && state.message == null) {
```
Comment (mickmister) [inline webapp/src/components/add_issue/add_issue.jsx:53]:
```suggestion
        if (props.visible && !state.message && props.message !== state.message) {
```

File: webapp/src/components/add_issue/add_issue.jsx [hunk 23]
```diff
@@ -0,0 +1,332 @@
...
+            return {message: props.message};
+        }
+        if (!props.visible && (state.message != null || state.sendTo != null)) {
+            return {
+                message: null,
```
Comment (mickmister) [inline webapp/src/components/add_issue/add_issue.jsx:58]:
```suggestion
                message: '',
```

File: webapp/src/components/add_issue/add_issue.jsx [hunk 24]
```diff
@@ -0,0 +1,332 @@
...
+                    <div className='AddIssueBox__content'>
+                        <div className='todoplugin-issue'>
+                            {this.state.previewMarkdown ? (
+                                <div
+                                    className='todoplugin-input'
+                                    style={style.markdown}
+                                >
+                                    {PostUtils.messageHtmlToComponent(
+                                        PostUtils.formatText(this.state.message),
+                                    )}
```
Comment (mickmister) [inline webapp/src/components/add_issue/add_issue.jsx:139]: Should we utilize some of the options we can pass into these functions?

**Review by mickmister — CHANGES_REQUESTED** [review]
Great work @asaadmahmood! The implementation mostly looks good. I have some changes requested

File: .vscode/extensions.json [hunk 25]
```diff
@@ -0,0 +1,6 @@
...
+    "recommendations": [
+        "golang.go",
+        "dbaeumer.vscode-eslint"
+    ]
+}
```
[the remark this answers is not in the document]
Comment (asaadmahmood) [reply .vscode/extensions.json:6]: @mickmister We don't have that ignored in focalboard, and it helps with developers using it. So I'd recommend keeping it.

File: webapp/src/components/add_issue/index.js [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+import {bindActionCreators} from 'redux';
+
+import {add, autocompleteUsers, openAssigneeModal, removeAssignee} from 'actions';
+import {getMessage, getPostID, getAssignee, getCurrentTeamRoute, isAddCardVisible} from 'selectors';
+
+import AddIssue from './add_issue';
+
+const mapStateToProps = (state) => ({
+    visible: isAddCardVisible(state),
+    message: getMessage(state) + (getPostID(state) ? '\n[Permalink](' + getCurrentTeamRoute(state) + 'pl/' + getPostID(state) + ')' : ''),
```
[the remark this answers is not in the document]
Comment (mickmister) [reply webapp/src/components/add_issue/index.js]: @asaadmahmood I see you've committed this suggestion verbatim, which is causing some syntax errors making it so the plugin can't compile. This code needs to go above the return statement. Then the return object can reference the `message` variable in its values.

File: webapp/src/components/sidebar_right/sidebar_right.jsx [hunk 27]
```diff
@@ -84,16 +102,25 @@ export default class SidebarRight extends React.PureComponent {
...
+        document.removeEventListener('keydown', this.handleKeypress);
         this.props.actions.setVisible(false);
     }
 
+    handleKeypress = (e) => {
+        if (e.altKey && isKeyPressed(e, Constants.KeyCodes.A)) {
+            e.preventDefault();
+            this.props.actions.openAddCard('');
+        }
+    };
```
[the remark this answers is not in the document]
Comment (asaadmahmood) [reply webapp/src/components/sidebar_right/sidebar_right.jsx:122]: We can do that, but maybe its an overkill right now. We don't do that anywhere in the system right now I think, we just do the action the user initiated with the shortcut.

File: webapp/src/widget/menu.tsx [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+        // No need to do anything, as click bubbled up to MenuWrapper, which closes
```
[the remark this answers is not in the document]
Comment (asaadmahmood) [reply webapp/src/widget/menu.tsx:43]: That's how its supplied in focalboard, so was just following that component.

File: webapp/src/widget/menuItem.tsx [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+    text?: string,
+    shortcut?: string,
+}
+
+const MenuItem = (props: Props) => {
+    const {icon, shortcut, action, text} = props;
+
+    useEffect(() => {
+        function handleKeypress(e: {key: string}) {
+            if (e.key === shortcut) {
```
[the remark this answers is not in the document]
Comment (asaadmahmood) [reply webapp/src/widget/menuItem.tsx:24]: Nops.

**Comment by mickmister** [comment]
@asaadmahmood It looks like there are some build issues from CI. If you can fix those then I can re-review the PR :+1:

File: .gitignore [hunk 30]
```diff
@@ -1 +1,2 @@
 dist
+.DS_Store
```
Comment (mickmister) [inline .gitignore:2]: Ideally something like this is in your own local global gitignore https://stackoverflow.com/a/7335487/2584550

File: webapp/src/widget/todo_toast/todo_toast.tsx [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+                className={generateClassName(classNames)}
+                title={props.title}
+            >
+                <div>
+                    <CompassIcon icon={props.content.icon}/>
+                    <span>{props.content.message}</span>
+                    <button
+                        onClick={undoTodo}
+                        className='TodoToast__undo'
+                    >{'Undo'}</button>
```
Comment (mickmister) [inline webapp/src/widget/todo_toast/todo_toast.tsx:66]:
```suggestion
                    <button
                        onClick={undoTodo}
                        className='TodoToast__undo'
                    >
                        {'Undo'}
                    </button>
```

File: webapp/src/components/add_issue/add_issue.jsx [hunk 32]
```diff
@@ -0,0 +1,328 @@
...
+            attachToThread: false,
+            previewMarkdown: false,
+            assigneeModal: false,
+        };
+    }
+
+    static getDerivedStateFromProps(props, state) {
+        if (props.visible && !state.message && props.message !== state.message) {
+            return {message: props.message};
+        }
```
Comment (mickmister) [inline webapp/src/components/add_issue/add_issue.jsx:51]: I'm wondering if this would cause an issue with:

- I'm editing the todo message
- I remove all text
- I do _some_ action that causes `getDerivedStateFromProps` to run (some prop or state value changes like `description`)
- The message has been changed from empty to `props.message`

Why not set this in the constructor instead?

File: webapp/src/components/add_issue/add_issue.jsx [hunk 33]
```diff
@@ -0,0 +1,328 @@
...
+        autocompleteUsers: PropTypes.func.isRequired,
+        openAssigneeModal: PropTypes.func.isRequired,
+        removeAssignee: PropTypes.func.isRequired,
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            message: '',
```
Comment (mickmister) [inline webapp/src/components/add_issue/add_issue.jsx:39]: Not sure if this messes with the flow, but it seems we should be initializing like this

```suggestion
            message: props.message || '',
```

File: webapp/src/components/todo_issues/todo_issues.jsx [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+function ToDoIssues(props) {
+    const style = getStyle(props.theme);
+    const {theme, siteURL, accept, complete, list, remove, bump, addVisible, issues} = props;
+
+    let emptyState = (
+        <div style={style.completed.container}>
+            <Tada/>
+            <h3 style={style.completed.title}>{'All tasks completed'}</h3>
+            <p style={style.completed.subtitle}>
+                {'Nicely done, you’ve finished all of your tasks! Why not reward yourself with a little break.'}
```
Comment (mickmister) [inline webapp/src/components/todo_issues/todo_issues.jsx:24]:
```suggestion
                {'Nicely done, you’ve finished all of your tasks! Why not reward yourself with a little break?'}
```

**Review by mickmister — APPROVED** [review]
LGTM :+1:

FYI I have this version running at https://assad-todo-updates.test.mattermost.cloud

**Comment by DHaussermann** [comment]
@asaadmahmood and @mickmister I can start testing on this shortly. But, it might make sense to get the `v0.6.0` release with App Bar support bout first since we've already done release testing.

**Comment by asaadmahmood** [comment]
Okay, cc @matthewbirtch @aaronrothschild

**Comment by matthewbirtch** [comment]
> @asaadmahmood and @mickmister I can start testing on this shortly. But, it might make sense to get the `v0.6.0` release with App Bar support out first since we've already done release testing.

@DHaussermann I don't see a reason this work couldn't be tested as-is. The work @asaadmahmood did has nothing to do with the app bar support work

File: webapp/src/components/add_issue/add_issue.jsx (hunk 32 above)
Comment (mickmister) [reply webapp/src/components/add_issue/add_issue.jsx:51]: @asaadmahmood Testing the PR out, this bug is present in the new code. Repro steps:

- From the post actions menu, click `Add Todo`
- Note that the todo message is pre-populated
- Erase all of the text of the message

Expected: The message field is empty

Observed: The message field's contents reverts to the original value of the post's message

**Comment by andrewbrown00** [comment]
Hey @DHaussermann, are you able to get this QA'd soon?

**Comment by DHaussermann** [comment]
Hey @andrewbrown00 and @asaadmahmood @malik took a look at this. I also took a look over this morning and this is looking good 👍 

I did find one issue but it may be a modal issue not specific to ToDo. So, I documented the issue here https://github.com/mattermost/mattermost-plugin-todo/issues/173

I'd be okay merging this now and investigating the issue separately unless either of you would prefer to add to this PR. Please advise.

**Comment by asaadmahmood** [comment]
@DHaussermann If this looks good, I'd like to get it in, that particular issue can be fixed as a separate PR, and we may also find other improvements when it gets to community, so those can be done as well.

**Review by DHaussermann — APPROVED** [review]
Tested and passed. 
- These enhancement provide a much needed improvement 
- UX enhancements have been reviewed and approved by QA
- Functional testing done to verify the details provided by Asaad
- 1 issue was found that can be addressed separately

LGTM! 
Huge thanks for this @asaadmahmood 🎉 Sorry for the delay in testing. Please merge.
